crates/valence_new/src/inventory.rs

```diff
@@ -425,3 +445,133 @@ pub(crate) fn update_client_on_close_inventory(
         }
     }
 }
+
+pub(crate) fn handle_click_container(
+    mut clients: Query<(&mut Client, &mut Inventory, Option<&mut OpenInventory>)>,
+    mut inventories: Query<&mut Inventory, Without<Client>>,
+    mut events: EventReader<ClickContainer>,
+) {
+    for event in events.iter() {
+        if event.window_id == 0 {
+            // the client is interacting with their own inventory
+            match clients.get_mut(event.client.clone()) {
+                Ok((mut client, mut inventory, _)) => {
+                    if client.inventory_state_id.0 != event.state_id {
+                        // client is out of sync, resync, and ignore the click
+                        inventory.modified = u64::MAX;
+                        continue;
+                    }
+
+                    // TODO: do more validation on the click
+                    client.cursor_item = event.carried_item.clone();
+                    for slot_change in event.slot_changes.clone() {
+                        if (0i16..inventory.slot_count() as i16).contains(&slot_change.0) {
+                            inventory.replace_slot(slot_change.0 as u16, slot_change.1);
+                            client.inventory_slots_modified |= 1 << slot_change.0;
+                        } else {
+                            // the client is trying to interact with a slot that does not exist,
+                            // ignore
+                            warn!(
+                                "Client attempted to interact with slot {} which does not exist",
+                                slot_change.0
+                            );
+                        }
+                    }
+                }
+                _ => {
+                    // the client or inventory does not exist, ignore
+                    warn!(
+                        "Client attempted to interact with their own
+    inventory, but it does not exist"
+                    );
+                }
+            }
+        } else {
+            // the player is interacting with an inventory that is open
+            let open_inventory = clients
+                .get_component::<OpenInventory>(event.client)
+                .ok()
+                .map(|open_inventory| open_inventory.entity);
+
+            if open_inventory.is_none() {
+                // the client is not viewing an inventory, ignore
+                warn!(
+                    "Client attempted to interact with an inventory, but is
+    not viewing one"
+                );
+                continue;
+            }
+
+            if let Ok(mut target_inventory) =
+                inventories.get_component_mut::<Inventory>(open_inventory.unwrap())
```

Comment:
I've been avoiding let-else because rustfmt still [hasn't decided how they want to format it](https://github.com/rust-lang/rustfmt/issues/4914), but I'll change it if that's what you want.